File: MarketplaceJPA/src/main/java/com/skilldistillery/marketplace/entities/Message.java
package com.skilldistillery.marketplace.entities;

import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;

@Entity
public class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    private String content;

    @CreationTimestamp
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @ManyToOne
    @JoinColumn(name = "sender_id")
    private User sender;

    @ManyToOne
    @JoinColumn(name = "recipient_id")
    private User recipient;

    @OneToMany(mappedBy = "message")
    private List<Message> replies;

    @ManyToOne
    @JoinColumn(name = "in_reply_to")
    private Message message;

    @Override
    public String toString() {
        return "Message [id=" + id + ", content=" + content + ", createdAt=" + createdAt + ", sender=" + sender
                + ", recipient=" + recipient + ", replies=" + replies + ", message=" + message + "]";
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public User getSender() {
        return sender;
    }

    public void setSender(User sender) {
        this.sender = sender;
    }

    public User getRecipient() {
        return recipient;
    }

    public void setRecipient(User recipient) {
        this.recipient = recipient;
    }

    public List<Message> getReplies() {
        return replies;
    }

    public void setReplies(List<Message> replies) {
        this.replies = replies;
    }

    public Message getMessage() {
        return message;
    }

    public void setMessage(Message message) {
        this.message = message;
    }

}

File: MarketplaceJPA/src/main/java/com/skilldistillery/marketplace/entities/Message.kt
package com.skilldistillery.marketplace.entities

import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime
import javax.persistence.*

@Entity
class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id = 0

    @Column(name = "content")
    var content: String? = null

    @CreationTimestamp
    @Column(name = "created_on")
    var createdOn: LocalDateTime? = null

    @UpdateTimestamp
    @Column(name = "updated_on")
    var updatedOn: LocalDateTime? = null

    @ManyToOne
    @JoinColumn(name = "sender_id")
    var sender: User? = null

    @ManyToOne
    @JoinColumn(name = "recipient_id")
    var recipient: User? = null

    @OneToMany(mappedBy = "message")
    var replies: List<Message>? = null

    @ManyToOne
    @JoinColumn(name = "in_reply_to")
    var message: Message? = null
    override fun toString(): String {
        return "Message{" +
                "id=" + id +
                ", content='" + content + '\'' +
                ", createdOn=" + createdOn +
                ", updatedOn=" + updatedOn +
                ", sender=" + sender +
                ", recipient=" + recipient +
                ", replies=" + replies +
                ", message=" + message +
                '}'
    }
}
